### PR TITLE
[site-isolation] Pass <iframe> 'scrolling' attribute to remote frame processes

### PR DESCRIPTION
**Pull request by mattwoodrow**
#### 098e3b169a3062dc2a4061c4e915d4a01f599f30
<pre>
[site-isolation] Pass &lt;iframe&gt; &apos;scrolling&apos; attribute to remote frame processes
<a href="https://bugs.webkit.org/show_bug.cgi?id=281103">https://bugs.webkit.org/show_bug.cgi?id=281103</a>
&lt;<a href="https://rdar.apple.com/137560506">rdar://137560506</a>&gt;

Reviewed by Said Abou-Hallawa.

Implements updateScrollingMode for RemoteFrame, sending it to the corresponding
LocalFrame via the UI process.

Two existing tests duplicated to be cross-origin tests for the initial and
mutation cases.

* LayoutTests/http/tests/site-isolation/compositing/iframes/iframe-content-flipping-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/compositing/iframes/iframe-content-flipping.html: Added.
* LayoutTests/http/tests/site-isolation/compositing/iframes/iframe-scrolling-change-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/compositing/iframes/iframe-scrolling-change.html: Added.
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::setOwnerElement):
* Source/WebCore/page/Frame.h:
(WebCore::Frame::updateScrollingMode): Deleted.
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::LocalFrame):
(WebCore::LocalFrame::createMainFrame):
(WebCore::LocalFrame::createSubframe):
(WebCore::LocalFrame::createProvisionalSubframe):
(WebCore::LocalFrame::createView):
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::updateScrollingMode):
* Source/WebCore/page/RemoteFrame.h:
* Source/WebCore/page/RemoteFrameClient.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/FrameInfoData.h:
* Source/WebKit/Shared/FrameInfoData.serialization.in:
* Source/WebKit/Shared/ProvisionalFrameCreationParameters.h:
* Source/WebKit/Shared/ProvisionalFrameCreationParameters.serialization.in:
* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm:
(WebKit::UIDelegate::UIClient::callDisplayCapturePermissionDelegate):
(WebKit::UIDelegate::UIClient::decidePolicyForUserMediaPermissionRequest):
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::ProvisionalPageProxy):
(WebKit::ProvisionalPageProxy::cancel):
(WebKit::ProvisionalPageProxy::initializeWebPage):
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::WebFrameProxy):
(WebKit::WebFrameProxy::didCreateSubframe):
(WebKit::WebFrameProxy::updateScrollingMode):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::create):
(WebKit::WebFrameProxy::scrollingMode const):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::initializeWebPage):
(WebKit::WebPageProxy::didCreateSubframe):
(WebKit::WebPageProxy::updateScrollingMode):
(WebKit::WebPageProxy::decidePolicyForNavigationActionSync):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::navigationActionData const):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::updateScrollingMode):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::createSubframe):
(WebKit::WebFrame::info const):
(WebKit::WebFrame::createProvisionalFrame):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::updateFrameScrollingMode):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/285222@main">https://commits.webkit.org/285222@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/2823358c5c1ea8131edf344439260cf4e9b8c2c8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71748 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51161 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24522 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75863 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22953 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73863 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58962 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22773 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56644 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15135 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74814 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46412 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61805 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37122 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/71256 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43073 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19272 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21294 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64980 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19635 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77582 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15982 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18819 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64368 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16026 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61838 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64375 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15888 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12536 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6175 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46961 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1740 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48032 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49316 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47774 "Built successfully") | | |

### CODE REVIEW

**Comment by smfr** [comment]
Could you separate the attribute change handling from the site isolation changes please? I think you should use https://bugs.webkit.org/show_bug.cgi?id=24070 for the first part.

File: Source/WebCore/html/HTMLFrameElementBase.cpp [hunk 1]
```diff
@@ -118,6 +118,9 @@ void HTMLFrameElementBase::openURL(LockHistory lockHistory, LockBackForwardList
 
 void HTMLFrameElementBase::attributeChanged(const QualifiedName& name, const AtomString& oldValue, const AtomString& newValue, AttributeModificationReason attributeModificationReason)
 {
+    if (name == scrollingAttr && contentFrame())
```
Comment (shallawa) [inline Source/WebCore/html/HTMLFrameElementBase.cpp]: If `name == scrollingAttr`, why do not we make an early return? Do we need to fall back to `HTMLFrameOwnerElement::attributeChanged()`?

File: Source/WebCore/page/Frame.cpp [hunk 2]
```diff
@@ -278,6 +278,7 @@ void Frame::setOwnerElement(HTMLFrameOwnerElement* element)
     if (element) {
         element->clearContentFrame();
         element->setContentFrame(*this);
+        setScrollingMode(element->scrollingMode());
```
Comment (shallawa) [inline Source/WebCore/page/Frame.cpp]: This call is a little bit confusing. In this function we set `m_ownerElement = element;` So we do not have to pass `element->scrollingMode()` to the sub-classes because they can get it by calling `m_ownerElement->scrollingMode()`. So I would suggest making `Frame::setOwnerElement()` virtual.
```
void LocalFrame::setOwnerElement(HTMLFrameOwnerElement* element)
{
    Frame::setOwnerElement(element);
    if (element)
        m_scrollingMode = element->scrollingMode();
    if (RefPtr view = this->view())
        view->setCanHaveScrollbars(m_scrollingMode != ScrollbarMode::AlwaysOff);
}
```

File: Source/WebCore/page/Frame.h [hunk 3]
```diff
@@ -29,6 +29,7 @@
 #include "FrameTree.h"
 #include "OwnerPermissionsPolicyData.h"
 #include "PageIdentifier.h"
+#include "ScrollTypes.h"
```
Comment (shallawa) [inline Source/WebCore/page/Frame.h]: Should not we forward declare `ScrollbarMode` instead?

File: Source/WebCore/page/LocalFrame.cpp [hunk 4]
```diff
@@ -177,6 +177,8 @@ LocalFrame::LocalFrame(Page& page, ClientCreator&& clientCreator, FrameIdentifie
     frameCounter.increment();
 #endif
 
+    setScrollingMode(scrollingMode);
```
Comment (shallawa) [inline Source/WebCore/page/LocalFrame.cpp]: Can `scrollingMode` be different from `ownerElement->scrollingMode()`?

File: Source/WebCore/page/LocalFrame.cpp [hunk 5]
```diff
@@ -194,17 +196,17 @@ void LocalFrame::init()
 
 Ref<LocalFrame> LocalFrame::createMainFrame(Page& page, ClientCreator&& clientCreator, FrameIdentifier identifier, SandboxFlags effectiveSandboxFlags, Frame* opener)
 {
-    return adoptRef(*new LocalFrame(page, WTFMove(clientCreator), identifier, effectiveSandboxFlags, nullptr, nullptr, opener));
+    return adoptRef(*new LocalFrame(page, WTFMove(clientCreator), identifier, effectiveSandboxFlags, ScrollbarMode::Auto, nullptr, nullptr, opener));
```
Comment (shallawa) [inline Source/WebCore/page/LocalFrame.cpp:200]: Oh I see the `ownerElement` can be nullptr. So I would suggest passing `std::optional<ScrollbarMode>` to the `LocalFrame` constructor which can make:
```
ASSERT(scrollbarMode ^ ownerElement);
scrollbarMode.value_or(ownerElement->scrollbarMode);
```

File: Source/WebCore/page/RemoteFrame.cpp [hunk 6]
```diff
@@ -161,4 +161,10 @@ OptionSet<AdvancedPrivacyProtections> RemoteFrame::advancedPrivacyProtections()
     return m_advancedPrivacyProtections;
 }
 
+void RemoteFrame::setScrollingMode(ScrollbarMode scrollingMode)
+{
+    ASSERT(ownerElement());
```
Comment (shallawa) [inline Source/WebCore/page/RemoteFrame.cpp]: I am not sure why we assert here ownerElement() is not null even it is not used.

File: Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp [hunk 7]
```diff
@@ -105,6 +107,7 @@ std::optional<NavigationActionData> WebFrameLoaderClient::navigationActionData(c
         WTFMove(originatingFrameID),
         WTFMove(parentFrameID),
         getCurrentProcessID(),
+        coreLocalFrame ? coreLocalFrame->scrollingMode() : ScrollbarMode::Auto,
```
Comment (shallawa) [inline Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp]: I am not sure why we get `coreLocalFrame()` instead of `coreFrame()`? And why we use ScrollbarMode::Auto if the frame is remote?

File: Source/WebKit/WebProcess/WebPage/WebPage.cpp [hunk 8]
```diff
@@ -3384,6 +3384,44 @@ void WebPage::updateFrameSize(WebCore::FrameIdentifier frameID, WebCore::IntSize
...
+{
+    if (!m_page)
+        return;
+
+    ASSERT(m_page->settings().siteIsolationEnabled());
+    RefPtr webFrame = WebProcess::singleton().webFrame(frameID);
+    if (!webFrame)
+        return;
+
+    RefPtr frame = webFrame->coreLocalFrame();
```
Comment (shallawa) [inline Source/WebKit/WebProcess/WebPage/WebPage.cpp]: This is another hard to see WebProcess -> WebProcess through UIProcess IPC.

File: Source/WebCore/page/RemoteFrame.cpp (hunk 6 above)
Comment (mattwoodrow) [reply Source/WebCore/page/RemoteFrame.cpp]: Admittedly confusing :)

The idea is that the only code that is allowed to set the scrolling mode on a RemoteFrame, is the process that has the outer document (so ownerElement() is non-null).

Every other WebProcess that is involved in the frame tree will also have a RemoteFrame instance for the same FrameID (except for the process where the frame actually lives, which has a LocalFrame), but setting the scrolling mode from those wouldn't make sense.

File: Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp (hunk 7 above)
Comment (mattwoodrow) [reply Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp]: `coreLocalFrame()` is the only one that has the `scrollingMode() ` getter.

As to the fallback, I'm not actually sure. We really do need a value, not sure what to do if the frame isn't local.

File: Source/WebKit/WebProcess/WebPage/WebPage.cpp (hunk 8 above)
Comment (mattwoodrow) [reply Source/WebKit/WebProcess/WebPage/WebPage.cpp]: Unfortunately there will be many many more :(

File: Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp (hunk 7 above)
Comment (mattwoodrow) [reply Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp]: @achristensen07 any idea how we should handle this path?

Comment (achristensen07) [reply Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp]: I'm not sure.  Either having a default value or having RemoteFrames know their value and broadcast updates to all processes.  I don't know this code well enough to say which is appropriate in this case.  It's also something that we might be able to handle in a separate PR.

Comment (mattwoodrow) [reply Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp]: I've fixed this by rebasing on top of #34999.

`decidePolicyForNavigationActionSync` no longer needs to implicitly create the frame, so we don't need to pass the scrolling mode in `NavigationActionData` at all.

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 9]
```diff
@@ -7759,10 +7769,10 @@ void WebPageProxy::decidePolicyForNavigationActionSync(IPC::Connection& connecti
...
     RefPtr frame = WebFrameProxy::webFrame(frameInfo.frameID);
     if (!frame) {
-        // This synchronous IPC message was processed before the asynchronous DidCreateSubframe one so we do not know about this frameID yet.
-        RefPtr parentFrame = WebFrameProxy::webFrame(frameInfo.parentFrameID);
-        MESSAGE_CHECK_BASE(parentFrame, connection);
-        parentFrame->didCreateSubframe(*frameInfo.frameID, frameInfo.frameName, data.effectiveSandboxFlags);
+        // This message should always be queued at this point, so we can pull it out with a 0 timeout.
+        connection.waitForAndDispatchImmediately<Messages::WebPageProxy::DidCreateSubframe>(webPageIDInMainFrameProcess(), 0_s);
+        frame = WebFrameProxy::webFrame(frameInfo.frameID);
+        MESSAGE_CHECK_COMPLETION_BASE(frame, connection, reply({ }));
```
Comment (shallawa) [inline Source/WebKit/UIProcess/WebPageProxy.cpp]: Is this change related to this PR? It looks like you are now processing the asynchronous message `WebPageProxy::DidCreateSubframe` instead of explicitly calling `didCreateSubframe()`. right?

**Review by shallawa — APPROVED** [review]

File: Source/WebKit/UIProcess/WebPageProxy.cpp (hunk 9 above)
Comment (mattwoodrow) [reply Source/WebKit/UIProcess/WebPageProxy.cpp]: Related, yes, but it's part of a separate PR.

Github unfortunately doesn't make this very visible. Will rebase now so that it goes away.
